Dockerfile.moc-xdmod

```diff
@@ -0,0 +1,100 @@
+# for a different zone.
+RUN sed -i 's/.*date.timezone[[:space:]]*=.*/date.timezone = UTC/' /etc/php.ini && \
+    sed -i 's/.*memory_limit[[:space:]]*=.*/memory_limit = -1/' /etc/php.ini && \
+    rm /etc/localtime && \
+    ln -s /usr/share/zoneinfo/UTC /etc/localtime
+
+# Setup Postfix
+RUN sed -ie 's/inet_interfaces = localhost/#inet_interfaces = localhost/' /etc/postfix/main.cf  && \
+    sed -ie 's/smtp      inet  n       -       n       -       -       smtpd/#smtp      inet  n       -       n       -       -       smtpd/' /etc/postfix/master.cf && \
+    sed -ie 's/smtp      unix  -       -       n       -       -       smtp/smtp      unix  -       -       n       -       -       local/' /etc/postfix/master.cf && \
+    sed -ie 's/relay     unix  -       -       n       -       -       smtp/relay     unix  -       -       n       -       -       local/' /etc/postfix/master.cf && \
+    echo '/.*/ root' >> /etc/postfix/virtual && \
+    postmap /etc/postfix/virtual && \
+    echo 'virtual_alias_maps = regexp:/etc/postfix/virtual' >> /etc/postfix/main.cf && \
+    newaliases
+
+# Install Composer
+RUN php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" && \
+    EXPECTED_SIGNATURE="$(wget -q -O - https://composer.github.io/installer.sig)" && \
+    ACTUAL_SIGNATURE="$(php -r "echo hash_file('SHA384', 'composer-setup.php');")" && \
+    if [ "$EXPECTED_SIGNATURE" != "$ACTUAL_SIGNATURE" ]; then echo 'ERROR: Invalid composer signature'; exit 1; fi && \
+    php composer-setup.php --install-dir=/bin --filename=composer --1 && \
+    php -r "unlink('composer-setup.php');"
+
+WORKDIR /root
+
+# the python clients required to connect to openstack
+RUN mkdir -p /root/rpmbuild/RPMS/noarch \
+    && mkdir /tmp/supremm \
+    && pip3 install --upgrade pip \
+    && pip3 install keystoneauth python-keystoneclient python-novaclient python-cinderclient cryptography python-openstacksdk mysql-connector-python pexpect
+
+# RPM install of xdmod
+RUN yum -y install  https://github.com/ubccr/xdmod/releases/download/v10.0.0/xdmod-10.0.0-1.0.el7.noarch.rpm 
+
+#COPY ./xdmod_conf/httpd.conf /etc/httpd/conf/httpd.conf
+RUN rm -f /etc/httpd/conf.d/ssl.conf \
+    && rm -f /run/httpd/httpd.pid
+
+COPY ./hypervisor_facts.py /usr/bin/xdmod-openstack-hypervisor
+COPY ./moc_openstack_api_reporting.py /usr/bin/xdmod-openstack-reporting
+COPY ./xdmod_init.py /usr/bin/xdmod-init 
+RUN chmod 777 /usr/bin/xdmod-openstack-hypervisor \
+    && sed -i -e 's/\r$//' /usr/bin/xdmod-openstack-hypervisor \
+    && chmod 777 /usr/bin/xdmod-openstack-reporting \
+    && sed -i -e 's/\r$//' /usr/bin/xdmod-openstack-reporting \ 
+    && chmod 777 /usr/bin/xdmod-init \
+    && sed -i -e 's/\r$//' /usr/bin/xdmod-init \ 
+    && chmod -R 777 /usr/bin \
+    && chmod -R 777 /var/log \
+    && chmod -R 777 /var/lib \
+    && chmod -R 777 /root \
+    && chmod -R 777 /run/httpd \
+    && chmod -R 777 /etc/xdmod \
+    && chmod -R 777 /etc/httpd 
```

Comment:
It really shouldn't be necessary to open up access like this on so many directories. We ought to be able to solve these with more granular permissions, and possibly the correct use of volumes for things that need to be writable like `/var/log` and `/run` (the latter of which could be a `tmpfs` mount).

Reply:
I was able to change the /run/httpd to just that directory.

Keep in mind, we are running this as a user process and not as root (as is expected by the xdmod team).   So here is a run down of the all of the permissions that you are mentioning:

        chmod -R 777 /usr/bin      # stores the xdmod executables (xdmod-setup, xdmid-shredder ... )
        chmod 777 /usr/bin         # This gave the error that xdmod-setup could not be run

        chmod -R 777 /var/log      # this needs to store log files written by a user process of apache, php and xdmod
        chmod -R 777 /var/lib      # this is used to store the session cookies used by xdmod

        chmod -R 777 /root         # This stores the data
        default                    # unable to create the subdirectories for the data        

        chmod 777 /run/httpd       # This is the directory of the pid file (removed the -R) 

        chmod -R 777 /etc/xdmod     # stores the configuration of xdmod
        default                     #  Was able to copy, but not able to modify  

        chmod -R 777 /etc/httpd  # stores all of the configuration for the apache    
        default                     # Was able to copy, but not able to modify

What permissions would be restrictive enough for you?  Are there a particular set you want me to test?

Reply:
@larsks - do you have an opinion on this?

Reply:
Yes, and sorry for the slow reply.

Looking at the deployment, we're mounting the following volumes:

```
- name: vol-xdmod-conf
  mountPath: /etc/xdmod
- name: vol-xdmod-src
  mountPath: /usr/share/xdmod
- name: vol-xdmod-data
  mountPath: /root/xdmod_data
- name: vol-httpd-conf
  mountPath: /etc/httpd
- name: vol-clouds-yaml
  mountPath: /etc/openstack
- name: vol-httpd-conf
  mountPath: /root/httpd
```

The user under which the container is running will already have
write access to all of these directories, although you will not be
able to access directories under `root`. We can demonstrate both facts
with the following example:

```
apiVersion: apps/v1

kind: Deployment
metadata:
  name: permission-test
spec:
  template:
    spec:
      containers:
        - image: docker.io/alpine:latest
          name: permission-test
          volumeMounts:
            - name: vol-xdmod-conf
              mountPath: /etc/xdmod
            - name: vol-xdmod-src
              mountPath: /usr/share/xdmod
            - name: vol-xdmod-data
              mountPath: /root/xdmod_data
          command:
            - sh
            - -c
            - |
              set -e

              for dir in /etc/xdmod /usr/share/xdmod /root/xdmod_data; do
                cd $dir
                pwd
                date > testfile
              done

              sleep inf
      volumes:
        - name: vol-xdmod-conf
          persistentVolumeClaim:
            claimName: pvc-xdmod-conf
        - name: vol-xdmod-src
          persistentVolumeClaim:
            claimName: pvc-xdmod-src
        - name: vol-xdmod-data
          persistentVolumeClaim:
            claimName: pvc-xdmod-data
```

The pod from that deployment will output:

```
/etc/xdmod
/usr/share/xdmod
sh: cd: line 3: can't cd to /root/xdmod_data: Permission denied
```

That is, it demonstrably is able to write files to `/etc/xdmod` and
`/usr/share/xdmod` without making any permissions changes. It is
unable to access `/root/xdmod_data` because the permissions on the
`/root` directory are:

```
$ ls -ld /root
drwx------    1 root     root            24 Jun 10 16:04 /root
```

But that doesn't mean we should change the permissions on `/root`:
because we *know* we're going to be running as a non-`root` user, we
should be using directories in a different location (which could be
pretty much anything: `/data`, `/home/xdmod`, whatever you want).

For directories like `/var/log/xdmod` and `/var/lib/<whatever>`, a
common solution would be to mount an `emptyDir` volume at both those
locations.  E.g., if we were to modify the above example `Deployment`,
we would add the following to the `volumes` section:

```
      volumes:
        - name: xdmod-state
          emptyDir: {}
```


And then the following to the `volumeMounts` section in our
containers:

```
          volumeMounts:
            - name: xdmod-state
              mountPath: /var/log/xdmod
            - name: xdmod-state
              mountPath: /var/lib/xdmod
```

With that volume mounted, while we cannot write to `/var/log` or
`/var/lib`:

```
~ $ touch /var/log/testfile
touch: /var/log/testfile: Permission denied
~ $ touch /var/lib/testfile
touch: /var/lib/testfile: Permission denied
```

We can write to `/var/log/xdmod` and `/var/lib/xdmod` without a
problem:

```
~ $ touch /var/log/xdmod/testfile
~ $ touch /var/lib/xdmod/testfile
```

I would not expect anything to be *writing* to `/usr/bin` at runtime,
so I'm not sure what's going on there.

Reply:
...and since you're also mounting a volume on `/etc/httpd`, that directory should also be writable by the current user without any permission changes.

Reply:
Do we really need a mount a separate disk to hold the pid file as we already have ephemeral storage (as in the container once it is deployed)?